### PR TITLE
Accept either string or list of strings for ScramArch at spec level

### LINKED ISSUE #7578: Accept string and list of strings for ScramArch
We need to update the request specs in order to accept either a simple string architecture or a list of architectures.

Then we also have to update the plugins such that they can use a different requirements string for that.

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes #7578
Deeper changes at CMSSW level done here https://github.com/dmwm/WMCore/pull/7579

We then validate and convert the user input into a python list of strings (even if user provided only a basic string), otherwise it's a bit more complicated to do the data sanitisation.

Still to be tested.

### CODE REVIEW

**Comment by ericvaandering** [comment]
Looks like this is blowing up in Jenkins...

**Comment by ericvaandering** [comment]
WMCore_t.WMSpec_t.StdSpecs_t.MonteCarloFromGEN_t.MonteCarloFromGENTest.testMonteCarloFromGEN says scramArch is None

**Comment by amaltaro** [comment]
Com'on!! It's supposed to be a simple change :(

**Comment by amaltaro** [comment]
test this please

**Comment by ericvaandering** [comment]
Almost there. :-)

**Comment by amaltaro** [comment]
And it should get fixed with my last commit. I like it, it means we have a quite reasonable unit test coverage now (though there is a lot to improve, of course) :-)

**Comment by amaltaro** [comment]
Most of the StdBase pylint warnings were fixed in 7619. In terms of spec level, it works as expected. I also took a change to look at a couple of condor jdl and a job runtime and everything looks fine (though we have to check it again in the near future).

From my side it's good to go.

File: src/python/WMCore/WMSpec/WMWorkloadTools.py
Comment (ericvaandering) [inline src/python/WMCore/WMSpec/WMWorkloadTools.py]: I'd rather see this in Utils (perhaps true of everything here) with a generic exception. But I think there is already exactly this in WMObject if you are inheriting from that. 

Having this general purpose stuff scattered around is a PITA.

File: test/python/WMCore_t/WMSpec_t/StdSpecs_t/StepChain_t.py
Comment (ericvaandering) [inline test/python/WMCore_t/WMSpec_t/StdSpecs_t/StepChain_t.py]: Nicely done. :-)

File: test/python/WMCore_t/WMSpec_t/WMWorkloadTools_t.py
Comment (ericvaandering) [inline test/python/WMCore_t/WMSpec_t/WMWorkloadTools_t.py]: Since it has the same behavior, you should test that's true with a few of your other examples. Test make list against make non empty

**Review by ericvaandering — CHANGES_REQUESTED** [review]

File: src/python/WMCore/WMSpec/WMWorkloadTools.py
Comment (amaltaro) [inline src/python/WMCore/WMSpec/WMWorkloadTools.py]: Hmm, I'm not sure we can define (cast) the spec input parameters to a WMObject (if you have an idea, I'm all ears). Perhaps what we could do is to add these functions as WMObject methods and create some sort of wrapper functions in Utils (but then we have the Utils dependency on WMCore/DataStructs).

Or I could simply move these functions to Utils/Utilities or something like that.

File: test/python/WMCore_t/WMSpec_t/WMWorkloadTools_t.py
Comment (amaltaro) [inline test/python/WMCore_t/WMSpec_t/WMWorkloadTools_t.py]: will do

File: src/python/WMCore/WMSpec/WMWorkloadTools.py
Comment (ericvaandering) [inline src/python/WMCore/WMSpec/WMWorkloadTools.py]: Definitely no dependencies from Utils onto something else. The idea with Utils is that *any* python project might benefit just by grabbing it.

What would it take to just move both make*List functions to Utils?

**Comment by ticoann** [comment]
test this please

**Comment by amaltaro** [comment]
Eric, I changed the unit tests as you suggested. Can you check and approve it if it's Ok?
As we discussed in the meeting, I'll work on this issue https://github.com/dmwm/WMCore/issues/7633 once this one gets merged.

**Review by ericvaandering — APPROVED** [review]

**Comment by amaltaro** [comment]
To be on the safe side, triggering jenkins once again.

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
@ticoann I think we can merge it now

**Review by ticoann — APPROVED** [review]

**Pull request merged**
